### PR TITLE
[Doc] Keccak Doc

### PR DESCRIPTION
**Pull request by nyunyunyunyu**

### CODE REVIEW

File: hashes/zkevm/src/keccak/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# ZKEVM Keccka
```
Comment (jonathanpwang) [inline hashes/zkevm/src/keccak/README.md:1]: typo

File: hashes/zkevm/src/keccak/README.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+## Vanilla
+Keccak circuit in vanilla halo2. This implementation starts from Scroll's version then adopts some changes from [this PR](https://github.com/scroll-tech/zkevm-circuits/pull/216) and [PSE version](https://github.com/privacy-scaling-explorations/zkevm-circuits/tree/main/zkevm-circuits/src/keccak_circuit).
+
+The majoir differences is that this version directly represent raw inputs and Keccak results as witnesses, while the original version only has RLCs(random linear combination) of raw inputs and Keccak results. Becuase this version doesn't need RLCs, it doesn't have the 2nd phase or use challenage APIs.
```
Comment (jonathanpwang) [inline hashes/zkevm/src/keccak/README.md:5]: typo; please check for typos throughout.

File: hashes/zkevm/src/keccak/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# ZKEVM Keccka
+## Vanilla
+Keccak circuit in vanilla halo2. This implementation starts from Scroll's version then adopts some changes from [this PR](https://github.com/scroll-tech/zkevm-circuits/pull/216) and [PSE version](https://github.com/privacy-scaling-explorations/zkevm-circuits/tree/main/zkevm-circuits/src/keccak_circuit).
```
Comment (jonathanpwang) [inline hashes/zkevm/src/keccak/README.md:3]: this did _not_ start from Scroll's version. Scroll's version is also based on the PSE version, which is due to Brecht. I had the correct attribution here: https://github.com/axiom-crypto/halo2-lib/blob/70c5cc05f8e7e0904533ae4044191dcd16bea214/hashes/zkevm-keccak/src/lib.rs#L2

File: hashes/zkevm/src/keccak/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+### Change Details
+- Removed column `input_rlc`/`input_len` and related gates.
+- Removed column `output_rlc` and related gates.
+- Removed challenges.
+- Refactored the folder strcuture to follow [Scroll's repo](https://github.com/scroll-tech/zkevm-circuits/tree/95f82762cfec46140d6866c34a420ee1fc1e27c7/zkevm-circuits/src/keccak_circuit). `mod.rs` and `witness.rs` could be found [here](https://github.com/scroll-tech/zkevm-circuits/blob/develop/zkevm-circuits/src/keccak_circuit.rs). `KeccakTable` could be found [here](https://github.com/scroll-tech/zkevm-circuits/blob/95f82762cfec46140d6866c34a420ee1fc1e27c7/zkevm-circuits/src/table.rs#L1308).
+- Imported utilites from [PSE zkevm-circuits repo](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/588b8b8c55bf639fc5cbf7eae575da922ea7f1fd/zkevm-circuits/src/util/word.rs). 
+
+## Coprocessor
+Keccak coprocessor circuits and utilities based on Halo2lib.
```
Comment (jonathanpwang) [inline hashes/zkevm/src/keccak/README.md:44]: `halo2-lib`

File: hashes/zkevm/src/keccak/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+### Motivation
+Move expensive Keccak computation into standalone circuits(**Coprocessor Circuits**) and circuits with actual business logic(**App Circuits**) can read Keccak results from coprocessor circuits. Then we have better scalability the maximum size of a single circuit could be managed and coprocessor/app circuits could be proved in paralle.
+
+### Output
+Logically a coprocessor ciruit outputs 3 columns `lookup_key`, `hash_lo`, `hash_hi` with `capacity` rows, where `capacity` is a configurable parameter and it means the maximum number of keccak_f this circuit can perform.
+
+- `lookup_key` can be cheaply derived from a bytes input. Specs can be found at `keccak::coprocessor::encode::encode_native_input`. Also `keccak::coprocessor::encode` provides some utilities to encode bytes inputs in halo2lib.
```
Comment (jonathanpwang) [inline hashes/zkevm/src/keccak/README.md:52]: `halo2-lib`

**Review by jonathanpwang — CHANGES_REQUESTED** [review]
Please check for typos throughout.

Also please say a bit more / give an example of the vanilla table format if we only look at the first row per round. (I think having an actual markdown table could be very helpful visually - maybe GPT can just make this.)

And then same example but explaining what happens in the leaf circuit with encoding.

**Comment by nyunyunyunyu** [comment]
> Please check for typos throughout.
> 
> Also please say a bit more / give an example of the vanilla table format if we only look at the first row per round. (I think having an actual markdown table could be very helpful visually - maybe GPT can just make this.)
> 
> And then same example but explaining what happens in the leaf circuit with encoding.

GPT is pretty useful! It helped me correct typos and generate the prototype of the table

**Review by jonathanpwang — COMMENTED** [review]
Very nice tables! I think this will be very helpful.

One last question: is `is_final` actually 1 on the first 17 rows in the table?
From
```
is_final: is_final_block && round == NUM_ROUNDS && row_idx == 0,
```
I actually thought it was only going to be nonzero on row 250.

**Comment by nyunyunyunyu** [comment]
> Very nice tables! I think this will be very helpful.
> 
> One last question: is `is_final` actually 1 on the first 17 rows in the table? From
> 
> ```
> is_final: is_final_block && round == NUM_ROUNDS && row_idx == 0,
> ```
> 
> I actually thought it was only going to be nonzero on row 250.

My understanding on is_final was inaccurate. Corrected the description.

**Pull request merged**